src/Microsoft.Azure.WebJobs.Host/Executors/FunctionExecutor.cs

```diff
@@ -175,6 +175,12 @@ private async Task<IDelayedException> TryExecuteAsyncCore(IFunctionInstanceEx fu
                 await HandleExceptionAsync(functionInstance.FunctionDescriptor.TimeoutAttribute, exceptionInfo, _exceptionHandler);
             }
 
+            if (parameterHelper.ReturnValue is RetryResult retry)
```

Comment:
The HandleException logic above ends up invoking the IWebJobsExceptionHandler, which in Functions we have an implementation that you should look at. Looks like your retry exceptions will pass through that handler and ultimately cause the Functions host to shut down.

I think for retries you need to handle them before and not let them flow through there, since they're not fatal exceptions.